File: mylang/go/internal/app/mylang/tid.go
package mylang

import (
	"mylang/internal/app/mylang/lex"
)

var TID = _MakeTID()

type _TIDType struct {
	cur       []*_Ident
	funcTable map[string]*_IdentFunc
	funcStack []string
}

func _MakeTID() _TIDType {
	return _TIDType{
		funcStack: []string{"global"},
		cur:       make([]*_Ident, 0),
		funcTable: make(map[string]*_IdentFunc),
	}
}

func (t *_TIDType) ICurTID(i int) *_Ident {
	return t.CurTID()[i]
}

func (t *_TIDType) CurTID() []*_Ident {
	return t.cur
}

func (t *_TIDType) Push(ident _Ident) {
	t.cur = append(t.cur, &ident)
}

func (t *_TIDType) Size() int {
	return len(t.cur)
}

func (t *_TIDType) DeclareFunc(ident _Ident, address int) bool {
	name := ident.Name()
	if _, found := t.funcTable[name]; found {
		return true
	}

	newIdent := _IdentFunc{_Ident: ident}
	newIdent.putValue(address)
	newIdent.putType(lex.FUNCTION)
	t.funcTable[name] = &newIdent
	return false
}

func (t *_TIDType) TopFunc() *_IdentFunc {
	return t.funcTable[t.funcStack[len(t.funcStack)-1]]
}

func (t *_TIDType) FindFunc(name string) *_IdentFunc {
	return t.funcTable[name]
}

func (t *_TIDType) PushFunc(name string) {
	t.funcStack = append(t.funcStack, name)
}

func (t *_TIDType) PopFunc() {
	t.funcStack = t.funcStack[:len(t.funcStack)-1]
}

func (t *_TIDType) TopFuncName() string {
	return t.funcStack[len(t.funcStack)-1]
}

func (t *_TIDType) PopBack() *_Ident {
	ident := t.cur[len(t.cur)-1]
	t.cur = t.cur[:len(t.cur)-1]
	return ident
}

func putIdent(buf string) int {
	prefix := TID.TopFuncName()
	idents := TID.CurTID()
	id := prefix + buf
	for i := range idents {
		if idents[i].ID() == id {
			return i
		}
	}

	newIdent := MakeIdent(buf)
	newIdent.setID(id)
	TID.Push(newIdent)
	return TID.Size() - 1
}

File: mylang/rust/src/tid.rs
use crate::ident::Ident;
use crate::ident_func::IdentFunc;
use crate::lex::Kind;
use std::collections::HashMap;
use std::ops::{Index, IndexMut};

pub struct TIDType {
	cur: Vec<Ident>,
	func_table: HashMap<String, crate::ident_func::IdentFunc>,
	func_stack: Vec<String>,
}

impl Index<usize> for TIDType {
	type Output = crate::ident::Ident;

	fn index(&self, index: usize) -> &Self::Output {
		&self.cur[index]
	}
}

impl IndexMut<usize> for TIDType {
	fn index_mut(&mut self, index: usize) -> &mut Self::Output {
		&mut self.cur[index]
	}
}

impl TIDType {
	pub fn new() -> TIDType {
		TIDType {
			cur: Vec::new(),
			func_table: HashMap::new(),
			func_stack: vec![String::from("global")],
		}
	}
	pub fn pop_back(&mut self) -> crate::ident::Ident {
		self.cur.pop().unwrap()
	}
	pub fn top_func_name(&self) -> &str {
		&self.func_stack[self.func_stack.len() - 1]
	}

	pub fn cur_mt(&mut self) -> &mut Vec<Ident> {
		&mut self.cur
	}

	pub fn cur(&self) -> &Vec<Ident> {
		&self.cur
	}

	pub fn put(&mut self, buf: &str) -> usize {
		let top_func_name = String::from(self.top_func_name());
		let idents = self.cur_mt();
		let mut id = String::from(top_func_name);
		id.push_str(buf);

		if let Some(index) = idents.iter().position(|r| r.id() == id) {
			return index;
		}

		let mut ident = Ident::new(String::from(buf));
		ident.set_id(id);
		idents.push(ident);
		return idents.len() - 1;
	}

	pub fn declare_func(&mut self, ident: &mut Ident, address: i32) -> bool {
		let name = ident.name();
		if self.func_table.contains_key(name) {
			return true;
		}
		let clone = ident.clone();
		let ident_func = IdentFunc::from_ident(&clone);
		ident.put_value(address);
		ident.put_kind(Kind::FUNCTION);
		let name = ident.name();
		self.func_table.insert(String::from(name), ident_func);
		return false;
	}

	pub fn push_func(&mut self, name: &str) {
		self.func_stack.push(String::from(name));
	}

	pub fn top_func(&mut self) -> &mut IdentFunc {
		let func: &str = &self.func_stack[self.func_stack.len() - 1];
		self.func_table.get_mut(func).unwrap()
	}

	pub fn len(&self) -> usize {
		self.cur().len()
	}

	pub fn pop_func(&mut self) {
		self.func_stack.pop();
	}

	pub fn find_func(&self, name: &str) -> std::option::Option<&IdentFunc> {
		self.func_table.get(name)
	}
}
